apps/aecore/src/aec_sync.erl

```diff
@@ -29,25 +45,142 @@
 connect_peer(Uri) ->
     gen_server:cast(?MODULE, {connect, Uri}).
 
+-spec received_block(aec_blocks:block()) -> ok.
+received_block(Block) ->
+    gen_server:cast(?MODULE, {received_block, Block}).
+
+block_created(Block) ->
+    gen_server:cast(?MODULE, {block_created, Block}).
+
+-spec subset_size() -> non_neg_integer().
+subset_size() ->
+    gen_server:call(?MODULE, subset_size).
+
+-spec set_subset_size(Sz :: non_neg_integer()) ->
+                             PrevSz when PrevSz :: non_neg_integer().
+set_subset_size(Sz) when is_integer(Sz), Sz > 0 ->
+    gen_server:call(?MODULE, {set_subset_size, Sz}).
+
+%% Builds a 'Ping' object for the initial ping.
+%% The 'Ping' object contains the following data:
+%% - source: our own peer uri
+%% - genesis_hash: base64-encoded hash of our genesis block
+%% - best_hash: base64-encoded hash of our top block
+%% - difficulty: (Ethereum: total_difficulty) our top_work
+%% - share: how many random peers we'd like the other node to share
+%% - peers: a random subset (size `Share`) that we know of
+%%
+%% Note that the peers element is not populated by this function.
+%% The pinger should insert a random list, excluding the pinged peer.
+%% The pinged should insert a random list, excluding the pinging peer,
+%%  as well as any peers that it shared.
+
+-type ping_object() :: map().
+-spec local_ping_object() -> ping_object().
+local_ping_object() ->
+    GHdr = aec_block_genesis:genesis_header(),
+    {ok, GHash} = aec_headers:hash_header(GHdr),
+    {ok, TopHdr} = aec_chain:top_header(),
+    {ok, TopHash} = aec_headers:hash_header(TopHdr),
+    Source = source_uri(),
+    {ok, {Difficulty, _}} = aec_chain:get_total_difficulty(),
+    #{<<"genesis_hash">> => base64:encode(GHash),
+      <<"best_hash">>    => base64:encode(TopHash),
+      <<"difficulty">>   => Difficulty,
+      <<"source">>       => Source,
+      <<"share">>        => 32,  % TODO: make this configurable
+      <<"peers">>        => []}.
+
+%% Compare local and remote 'Ping' objects.
+%% 1. Compare genesis blocks. If they differ, return error
+%% 2. Compare best_hash. If they are the same, we're in sync
```

Comment:
best hash - is it ethereum thing to name like that top hash?

Reply:
That was borrowed from ethereum, yes.